graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -869,6 +865,50 @@ def _validate_all_tags_are_used(metadata):
         )
 
 
+def _validate_output_types(outputs: Generator[Tuple[str, OutputInfo], None, None]):
+    """Ensure output types are valid.
+
+    Checks that _x_count output is within a fold scope, and that its type is GraphQLInt.
+    Checks that all other outputs within a fold scope have GraphQLList as their type.
+
+    Args:
+        outputs: generator producing (output names, OutputInfo) tuples describing a query's outputs
+
+    Raises:
+        AssertionError if an invalid output type is found
+
+    """
+    for name, info in outputs:
```

Comment:
It might be a bit more pleasant to pull this loop out of the function. I think the function would then a bit easier to reason about — no more generators in or out etc.